integrations/jenkins/client.py

```diff
@@ -0,0 +1,103 @@
+from typing import Any, AsyncGenerator, Optional
+from urllib.parse import urlparse
+import httpx
+
+from loguru import logger
+
+from port_ocean.context.event import event
+from port_ocean.utils import http_async_client
+
+
+class JenkinsClient:
+    def __init__(
+        self, jenkins_base_url: str, jenkins_user: str, jenkins_token: str
+    ) -> None:
+        self.jenkins_base_url = jenkins_base_url
+        self.client = http_async_client
+        self.client.auth = httpx.BasicAuth(jenkins_user, jenkins_token)
+
+    async def get_all_jobs_and_builds(
+        self,
+    ) -> AsyncGenerator[list[dict[str, Any]], None]:
+        cache_key = "jenkins_jobs_and_builds"
+
+        if cache := event.attributes.get(cache_key):
+            logger.info("picking from cache")
+            yield cache
+            return
+
+        all_jobs = []
+
+        try:
+            async for jobs in self.fetch_jobs_and_builds_from_api():
+                all_jobs.extend(jobs)
+
+            yield all_jobs
+            event.attributes[cache_key] = all_jobs
+        except Exception as e:
+            logger.error(f"Unexpected error occurred: {e}")
+            raise
+
+    async def fetch_jobs_and_builds_from_api(
```

Comment:
this function is way overloaded 
separate this pls
at the moment it is unreadable

you have recursions and loops and if checks way 2 much logic

Reply:
I am going to simplify the code and make it easier.

But I found recursion to be the best way to handle this scenario. Did you have an alternative way in mind?